File: packages/typescript-sdk/examples/cache-redis.ts
// @ts-nocheck
/**
 * Example: Custom Redis cache provider implementation
 *
 * This example demonstrates:
 * - Implementing the CacheProvider interface for Redis
 * - Using ioredis for Redis connections
 * - Configuring the client with a custom cache provider
 *
 * NOTE: This example requires Redis to be running and the ioredis package:
 *   npm install ioredis
 *   # or
 *   bun add ioredis
 *
 * Start Redis locally:
 *   docker run -d -p 6379:6379 redis:alpine
 *   # or
 *   brew install redis && brew services start redis
 */

import Redis from "ioredis";
import type { CacheProvider } from "../src/index.js";
import { BookName, LSBibleClient } from "../src/index.js";

/**
 * Redis cache provider implementation.
 *
 * Implements the CacheProvider interface to store cache entries in Redis
 * with automatic expiration using Redis TTL.
 */
class RedisCacheProvider implements CacheProvider {
  private redis: Redis; // Replace with: Redis

  constructor(redis: Redis /* Replace with: Redis */) {
    this.redis = redis;
  }

  async get<T>(key: string): Promise<T | undefined> {
    try {
      const value = await this.redis.get(key);
      return value ? JSON.parse(value) : undefined;
    } catch (error) {
      console.error(`Redis get error for key ${key}:`, error);
      return undefined;
    }
  }

  async set<T>(key: string, value: T, ttl: number): Promise<void> {
    try {
      // Store with expiration (EX = seconds)
      await this.redis.set(key, JSON.stringify(value), "EX", ttl);
    } catch (error) {
      console.error(`Redis set error for key ${key}:`, error);
    }
  }
}

async function main() {
  console.log("=== LSBible SDK - Redis Cache Example ===\n");

  // Initialize Redis client
  const redis = new Redis({
    host: process.env.REDIS_HOST || "localhost",
    port: Number(process.env.REDIS_PORT) || 6379,
    password: process.env.REDIS_PASSWORD,
    // Optional: configure connection retry strategy
    retryStrategy: (times) => {
      if (times > 3) {
        console.error("Redis connection failed after 3 retries");
        return null; // Stop retrying
      }
      return Math.min(times * 100, 2000); // Exponential backoff
    },
  });

  // Create cache provider
  const cacheProvider = new RedisCacheProvider(redis);

  // Initialize client with Redis cache
  const client = new LSBibleClient({
    cache: {
      provider: cacheProvider,
    },
  });

  try {
    console.log("1️⃣  Fetching John 3:16 (cache miss)...");
    const start1 = Date.now();
    const verse1 = await client.getVerse(BookName.JOHN, 3, 16);
    const duration1 = Date.now() - start1;
    console.log(`   ${verse1.title}: ${verse1.verses[0]?.plainText.substring(0, 50)}...`);
    console.log(`   ⏱️  Duration: ${duration1}ms\n`);

    console.log("2️⃣  Fetching John 3:16 again (cache hit)...");
    const start2 = Date.now();
    const verse2 = await client.getVerse(BookName.JOHN, 3, 16);
    const duration2 = Date.now() - start2;
    console.log(`   ${verse2.title}: ${verse2.verses[0]?.plainText.substring(0, 50)}...`);
    console.log(`   ⏱️  Duration: ${duration2}ms (${duration1 - duration2}ms faster!)\n`);

    console.log("3️⃣  Checking Redis cache...");
    const cacheKey = "verse:John 3:16";
    const ttl = await redis.ttl(cacheKey);
    console.log(`   Cache key: ${cacheKey}`);
    console.log(`   TTL remaining: ${ttl} seconds (~${Math.round(ttl / 86400)} days)\n`);

    console.log("✅ Done!");
  } catch (error) {
    if (error instanceof Error) {
      console.error(`❌ Error: ${error.message}`);
    } else {
      console.error("❌ Unknown error occurred");
    }
  } finally {
    // Clean up Redis connection
    await redis.quit();
  }

  console.log("📝 This example requires Redis and ioredis to be installed.");
  console.log("   See comments in the file for setup instructions.");
}

main();

File: packages/python-sdk/examples/cache_redis.py
"""Example: Custom Redis cache provider implementation

This example demonstrates:
- Implementing the CacheProvider protocol for Redis
- Using redis-py with async support
- Configuring the client with a custom cache provider

NOTE: This example requires Redis to be running and the redis package:
  pip install redis[asyncio]
  # or
  poetry add redis[asyncio]

Start Redis locally:
  docker run -d -p 6379:6379 redis:alpine
  # or
  brew install redis && brew services start redis
"""

import asyncio
import json
from typing import Any

# Uncomment after installing redis[asyncio]:
# import redis.asyncio as redis


class RedisCacheProvider:
    """Redis cache provider implementation.

    Implements the CacheProvider protocol to store cache entries in Redis
    with automatic expiration using Redis TTL.
    """

    def __init__(self, redis_client: Any):  # Replace Any with redis.Redis after installing
        self.redis = redis_client

    async def get(self, key: str) -> Any | None:
        try:
            value = await self.redis.get(key)
            return json.loads(value) if value else None
        except Exception as error:
            print(f"Redis get error for key {key}: {error}")
            return None

    async def set(self, key: str, value: Any, ttl: int) -> None:
        try:
            # Store with expiration (ex = seconds)
            await self.redis.set(key, json.dumps(value), ex=ttl)
        except Exception as error:
            print(f"Redis set error for key {key}: {error}")


async def main():
    print("=== LSBible SDK - Redis Cache Example ===\n")

    # NOTE: Uncomment after installing redis[asyncio] and starting Redis
    """
    # Initialize Redis client
    redis_client = redis.Redis(
        host=os.getenv("REDIS_HOST", "localhost"),
        port=int(os.getenv("REDIS_PORT", "6379")),
        password=os.getenv("REDIS_PASSWORD"),
        decode_responses=True,
        # Optional: configure connection retry strategy
        retry_on_timeout=True,
        socket_connect_timeout=5,
    )

    # Create cache provider
    cache_provider = RedisCacheProvider(redis_client)

    # Initialize client with Redis cache
    async with LSBibleClient(
        cache={"provider": cache_provider}
    ) as client:
        print("1️⃣  Fetching John 3:16 (cache miss)...")
        import time
        start1 = time.time()
        verse1 = await client.get_verse(BookName.JOHN, 3, 16)
        duration1 = (time.time() - start1) * 1000
        print(f"   {verse1.title}: {verse1.verses[0].plain_text[:50]}...")
        print(f"   ⏱️  Duration: {duration1:.0f}ms\n")

        print("2️⃣  Fetching John 3:16 again (cache hit)...")
        start2 = time.time()
        verse2 = await client.get_verse(BookName.JOHN, 3, 16)
        duration2 = (time.time() - start2) * 1000
        print(f"   {verse2.title}: {verse2.verses[0].plain_text[:50]}...")
        print(f"   ⏱️  Duration: {duration2:.0f}ms ({duration1 - duration2:.0f}ms faster!)\n")

        print("3️⃣  Checking Redis cache...")
        cache_key = "verse:John 3:16"
        ttl = await redis_client.ttl(cache_key)
        print(f"   Cache key: {cache_key}")
        print(f"   TTL remaining: {ttl} seconds (~{round(ttl / 86400)} days)\n")

        print("✅ Done!")

    # Clean up Redis connection
    await redis_client.close()
    """

    print("📝 This example requires Redis and redis[asyncio] to be installed.")
    print("   See comments in the file for setup instructions.")


if __name__ == "__main__":
    asyncio.run(main())
